### PR TITLE
add Node::getPathInHierarchy

### PR DESCRIPTION
**Pull request by novain1102**
Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/core/scene-graph/Node.cpp [hunk 1]
```diff
@@ -886,6 +886,16 @@ void Node::clearNodeArray() {
     }
 }
 
+ccstd::string Node::getPathInHierarchy() const {
+    ccstd::string result = getName();
+    Node *curNode = getParent();
+    while (curNode && dynamic_cast<Scene *>(curNode) == nullptr) {
```
Comment (novain1102) [inline native/cocos/core/scene-graph/Node.cpp]: In TS, curNode  is `BaseNode` Type ;`Scene` and `Node` is derived class of `BaseNode`

```ts
        let curNode: BaseNode | null = this.parent;
        while (curNode && curNode instanceof Node) {
           ...
        }
```

Comment (minggo) [inline native/cocos/core/scene-graph/Node.cpp]: dynamic_cast<Scene *>(curNode) == nullptr -> dynamic_cast<Scene *>(curNode) != nullptr

Comment (novain1102) [reply native/cocos/core/scene-graph/Node.cpp]: As I can see, the original idea in TS is to judge that curNode is not a Scene @Zhmz

File: native/cocos/core/scene-graph/Node.cpp [hunk 2]
```diff
@@ -886,6 +886,16 @@ void Node::clearNodeArray() {
...
+        result = ccstd::string(curNode->getName()).append("/").append(result);
```
Comment (minggo) [inline native/cocos/core/scene-graph/Node.cpp]: can reuse result, just insert contents ahead of it, such as
```c++
result.insert(0, "/").insert(0, curNode->getName());
```

Comment (novain1102) [reply native/cocos/core/scene-graph/Node.cpp]: It's a good idea!

File: native/cocos/core/scene-graph/Node.cpp (hunk 1 above)
Comment (minggo) [reply native/cocos/core/scene-graph/Node.cpp]: If so, then its parent is null, don't have to check type. Please double check the meaning of it.

Comment (Zhmz) [reply native/cocos/core/scene-graph/Node.cpp]: using "parent == nullptr" can exactly determine that the curNode is a scene type.
All comman nodes have their own parent but scenes don't have.

**Review by Zhmz — CHANGES_REQUESTED** [review]

File: native/cocos/core/scene-graph/Node.cpp (hunk 1 above)
Comment (novain1102) [reply native/cocos/core/scene-graph/Node.cpp]: > If so, then its parent is null, don't have to check type. Please double check the meaning of it.

Yes. You're right. Wait for @Zhmz to make sure.

File: native/cocos/core/scene-graph/Node.cpp [hunk 3]
```diff
@@ -886,6 +886,16 @@ void Node::clearNodeArray() {
     }
 }
 
+ccstd::string Node::getPathInHierarchy() const {
+    ccstd::string result = getName();
+    Node *curNode = getParent();
+    while (curNode && curNode->getParent()) {
```
Comment (minggo) [inline native/cocos/core/scene-graph/Node.cpp]: Don't have to check its parent. Current logic will exclude Scene.

Comment (minggo) [inline native/cocos/core/scene-graph/Node.cpp]: Don't have to check its parent. Current logic will exclude Scene.

**Review by minggo — APPROVED** [review]

File: native/cocos/core/scene-graph/Node.cpp (hunk 3 above)
Comment (minggo) [reply native/cocos/core/scene-graph/Node.cpp]: Sorry, it is different from TS. So should keep it.

**Pull request merged**